### ISSUE TITLE
rpi_ws281x: ws2811_init failed: -2 (Out of memory)

### ISSUE THREAD

**Issue by alde103** [GregMefford/blinkchain#20]
I use a rpi3b+ and after several amount of calls  (like 30 calls/s or more) to Blinchain.HAL, the c code appears to timeout:

```
10:48:51.165 [error] GenServer Blinkchain.HAL terminating
** (RuntimeError) timeout waiting for rpi_ws281x OS process to reply
    (blinkchain) lib/blinkchain/hal.ex:190: Blinkchain.HAL.receive_from_port/1
    (blinkchain) lib/blinkchain/hal.ex:105: Blinkchain.HAL.handle_call/3
    (stdlib) gen_server.erl:661: :gen_server.try_handle_call/4
    (stdlib) gen_server.erl:690: :gen_server.handle_msg/6
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
Last message (from #PID<0.5232.73>): :render
State: %Blinkchain.HAL.State{config: %Blinkchain.Config{canvas: %Blinkchain.Config.Canvas{height: 2, width: 6}, channel0: %Blinkchain.Config.Channel{arrangement: [%Blinkchain.Config.Strip{count: 6, direction: :right, origin: {0, 0}, spacing: 1}], brightness: 255, gamma: nil, invert: false, number: 0, pin: 18, type: :grb}, channel1: %Blinkchain.Config.Channel{arrangement: [%Blinkchain.Config.Strip{count: 6, direction: :right, origin: {0, 1}, spacing: 1}], brightness: 255, gamma: nil, invert: false, number: 1, pin: 13, type: :grb}}, port: #Port<0.143367>, subscriber: nil}
Client #PID<0.5232.73> is alive
```

Overtime, another error appears:

```
16:35:55.223 [error] Process Blinkchain.HAL (#PID<0.6784.73>) terminating
** (RuntimeError) rpi_ws281x OS process died with status: 1
    (blinkchain) lib/blinkchain/hal.ex:188: Blinkchain.HAL.receive_from_port/1
    (blinkchain) lib/blinkchain/hal.ex:109: Blinkchain.HAL.handle_info/2
    (stdlib) gen_server.erl:637: :gen_server.try_dispatch/4
    (stdlib) gen_server.erl:711: :gen_server.handle_msg/6
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
Initial Call: Blinkchain.HAL.init/1
Ancestors: [Blinkchain.Supervisor, #PID<0.6714.73>, #PID<0.6451.73>]
Message Queue Length: 1
Messages: [{#Port<0.143412>, {:data, {:eol, 'rpi_ws281x: ws2811_init failed: -2 (Out of memory)'}}}]
Links: [#PID<0.6774.73>]
Dictionary: []
Trapping Exits: false
Status: :running
Heap Size: 4185
Stack Size: 27
Reductions: 8517
```

I use this config:
```
config :blinkchain,
  canvas: {6, 2},
  channel0: [
    pin: 18,
    type: :grb,
    arrangement: [
      %{
        type: :strip,
        origin: {0, 0},
        count: 6,
        direction: :right
      }
    ]
  ],
  channel1: [
    pin: 13,
    type: :grb,
    arrangement: [
      %{
        type: :strip,
        origin: {0, 1},
        count: 6,
        direction: :right
      }
    ]
  ]
```

**Comment by GregMefford** [comment]
Since you mentioned that you're using an rpi3b+, I wonder if the problem is just that [this still isn't merged](https://github.com/GregMefford/blinkchain/pull/16). I forgot about it and only just recently got my own RPI 3 A+ / B+ hardware, so now I can test it out and see if I can reproduce / solve the problem.

**Comment by GregMefford** [comment]
Ok, so I am seeing the same issue after upgrading the rainbow example to use the latest Nerves systems, even on `rpi0`, so I don't think it's directly related to the RPi 3+ issue, but it's still possibly related to a missing RPi firmware update, and it looks like the `rpi_ws281x` project has some updates, so I'll give that a try and see if it helps.

**Comment by GregMefford** [comment]
TL;DR:
I think I have a fix - I just need to clean up a few lingering things and make a new release.

Boring details:
After messing around with a lot of hardware/software permutations, it looks like there was a change at some point in the past year or so where [DMA channel 10](https://github.com/GregMefford/blinkchain/blob/a9c6874ced62cb422cad0ae6be58926eb4f7bf9a/src/rpi_ws281x.c#L13) has started to be used for other things in the Rpi firmware. The older Rpi Zero (without wireless) doesn't exhibit this problem, so I suspect it's something related to the WiFi/BlueTooth hardware. I have locally made that configurable and tried channel 5 instead (which ). Unfortunately, I'm not aware of any documentation about which DMA channels are "safe" for this kind of use without conflicts with the firmware. I believe channel 5 _used_ to be safe and then later was used for something, causing us to move to channel 10... maybe they realized it was breaking existing projects and moved back 🤷‍♂
